nixos/modules/services/backup/borgbackup.nix

```diff
@@ -0,0 +1,510 @@
+{ config, lib, pkgs, ... }:
+
+with lib;
+
+let
+
+  isPath = x: builtins.substring 0 1 x == "/";
+ 
+  mkExcludeFile = cfg:
+    # Write each exclude pattern to a new line
+    pkgs.writeText "excludefile" (concatStringsSep "\n" cfg.exclude);
+
+  mkPathsArgs = cfg:
+    # Shell-escape each path and put spaces between them
+    concatMapStringsSep " " escapeShellArg cfg.paths;
+
+  mkKeepArgs = cfg:
+    # If cfg.prune.keep e.g. has a yearly attribute,
+    # its content is passed on as --keep-yearly
+    concatStringsSep " "
+      (mapAttrsToList (x: y: "--keep-${x}=${toString y}") cfg.prune.keep);
+
+  mkBackupScript = cfg: ''
+    on_exit()
+    {
+      exit_status=$?
+      # Reset the EXIT handler, or else we're called again on 'exit' below
+      trap - EXIT
+      ${cfg.postHook}
+      exit $exit_status
+    }
+    trap 'on_exit' INT TERM QUIT EXIT
+
+    ${cfg.preHook}
+  '' + optionalString cfg.doInit ''
+    # Run borg init if the repo doesn't exist yet
+    if ! borg list; then
+      borg init \
+        --encryption ${cfg.encryption.mode} \
+        ${cfg.extraInitArgs}
+      ${cfg.postInit}
+    fi
+  '' + ''
+    borg create \
```

Comment:
Doh. I just realized that to use this module, I need a way to add extra arguments to "borg create", that are only known at runtime. See https://github.com/borgbackup/borg/issues/3635#issuecomment-369044029.

Hm, an alternative would be to inspect the NixOS configuration to find and exclude autofs mountpoints. But ugh...

Reply:
I will use `$extraCreateArgs` as a bash variable, just like `$archiveName`. Would that be okay?

Reply:
So the NixOS option initialized $extraCreateArgs, and then the hook can prepend/append to it? Sounds good.